Implement K-means clustering

msluszniak opened the pull request.

polvalente commented on lib/scholar/cluster/k_means.ex:
```diff
@@ -0,0 +1,325 @@
+
+  ## Options
+
+    * `:num_clusters` - The number of clusters to form as well as the number of centroids to generate. Defaults to `8`.
```
I would make `num_clusters` required. Also, perhaps this doc could emphasize this is the `K` in `K-means`

msluszniak replied:
Yes, I've been thinking about requiring num_clusters. In sci-kit learn 8 is the default, but the required parameter also makes sense.

polvalente commented on lib/scholar/cluster/k_means.ex:
```diff
@@ -0,0 +1,325 @@
+      )
+
+    verify(x, opts)
+    x = x |> Nx.as_type({:f, 32})
```
Do we need to forcibly upcast here?

polvalente replied:
Actually, this could downcast if the type is already f64. I would try to avoid this `as_type` here

polvalente commented on lib/scholar/cluster/k_means.ex:
```diff
@@ -0,0 +1,325 @@
+
+    first_centroid_idx = Nx.random_uniform({num_runs}, 0, num_samples - 1, type: {:u, 32})
+
+    first_centroid = Nx.take(x, first_centroid_idx) |> Nx.reshape({:auto})
```
```suggestion
    first_centroid = Nx.take(x, first_centroid_idx) |> Nx.flatten()
```

polvalente replied:
there are similar occurences below as well

polvalente commented on lib/scholar/cluster/k_means.ex:
```diff
@@ -0,0 +1,325 @@
+    first_centroid = Nx.take(x, first_centroid_idx) |> Nx.reshape({:auto})
+
+    indices_centroids =
+      Nx.concatenate([
+        Nx.new_axis(Nx.reshape(Nx.iota({num_runs, num_features}, axis: 0), {:auto}), 0),
+        Nx.new_axis(Nx.broadcast(0, {num_runs * num_features}), 0),
+        Nx.new_axis(Nx.tile(Nx.iota({num_features}), [num_runs]), 0)
+      ])
```
Can't this be Nx.stack instead?

Also, if we can avoid the transpose it would be better

polvalente commented on lib/scholar/cluster/utils.ex:
```diff
@@ -0,0 +1,21 @@
+  import Scholar.Shared
+
+  @spec squared_euclidean(Nx.t(), Nx.t()) :: Nx.t()
+  defn squared_euclidean(x, y, opts \\ []) do
```
We already have this function under Scholar.Metrics.Distance

msluszniak replied:
But not with axes specification

polvalente replied:
You can add it there instead

msluszniak replied:
I guess that we should add this option to all the other metrics to keep API consistent then, but it's an idea.

polvalente replied:
I agree. Perhaps it's better that this is done in a separate PR first

josevalim replied:
@msluszniak please change `Scholar.Metrics.Distance` to support the `:axes` option. :)

josevalim commented on lib/scholar/cluster/k_means.ex:
```diff
@@ -0,0 +1,322 @@
+
+    * `:max_iterations` - Maximum number of iterations of the k-means algorithm for a single run. Defaults to `300`.
+
+    * `:num_runs` - Number of time the k-means algorithm will be run with different centroid seeds. The final results will be the best output of num_runs runs in terms of inertia. Defaults to `10`.
```
Sorry for the nitpick but can you please trim the descriptions so the lines do not exceed 100 chars?

josevalim commented on lib/scholar/cluster/k_means.ex:
```diff
@@ -0,0 +1,322 @@
+    end
+  end
+
+  #
```
```suggestion
```

josevalim commented on lib/scholar/cluster/k_means.ex:
```diff
@@ -0,0 +1,322 @@
+
+  #
+  @doc """
+  Makes predictions with the given model on inputs `x`. It returns a tensor with clusters corresponding to the input.
```
```suggestion
  Makes predictions with the given model on inputs `x`.

  It returns a tensor with clusters corresponding to the input.
```

josevalim commented on test/scholar/cluster/k_means_test.exs:
```diff
@@ -0,0 +1,152 @@
+defmodule Scholar.Cluster.KMeansTest do
+  use ExUnit.Case, async: true
+
+  test "Simple test without weights" do
```
```suggestion
  test "without weights" do
```

josevalim commented on lib/scholar/cluster/k_means.ex:
```diff
@@ -0,0 +1,322 @@
+defmodule Scholar.Cluster.KMeans do
+  @moduledoc """
+  K-Means algorithm
```
```suggestion
  K-Means algorithm.
```

josevalim reviewed (approved):
This looks great @msluszniak!

Do you think you can also add a notebook that shows how to use this API, including some plottings? :)

msluszniak commented:
> 

Sure, that's a great idea! :)

josevalim commented on lib/scholar/cluster/k_means.ex:
```diff
@@ -14,21 +14,28 @@ defmodule Scholar.Cluster.KMeans do
   ## Options
 
-    * `:num_clusters` - The number of clusters to form as well as the number of centroids to generate. Defaults to `8`.
+    * `:num_clusters` - The number of clusters to form as well as the number of centroids to generate. Required.
```
Btw, why did you make them required? If 8 is a random number, perhaps a better default is 2? Or does Scipy make them required?

polvalente replied:
I suggested making the number of clusters required mainly because since any choice we provide as default is arbitrary, I didn't see a reason for it being optional.

josevalim replied:
I think the least arbitrary is 2 but I am fine with it being required or following Scipy.

jonatanklosko commented on lib/scholar/cluster/k_means.ex:
```diff
@@ -0,0 +1,337 @@
+    end
+
+    if opts[:random_state] != nil do
+      :rand.seed(:exsss, opts[:random_state])
```
This will only impact the BinaryBackend and this happens at defn compile time, which we don't want. I think we need to wait for https://github.com/elixir-nx/nx/pull/846 to do this sort of things.

As for tests that rely on this, maybe we could use data points that guarantee convergence, regardless the initialization. And/or have some property tests if there are any we can reasonably have.

msluszniak replied:
The problem will be only with the labels parameter. Depending on the initialization of centroids the resultant labels might be i. ex. [1, 0, 1, 0], but also [0, 1, 0, 1].

jonatanklosko replied:
You can normalize he classes for comparison.

```elixir
defmodule Scholar.Utils do
  import Nx.Defn

  @doc """
  Returns labels that unambiguously identify the given grouping.

  This function effectively renumbers the classes, such that the
  first occurrences of each classes are in increasing order.
  
  ## Examples

      iex> Scholar.Utils.normalize_labels(Nx.tensor([1, 0, 1, 0]), num_classes: 2)
      #Nx.Tensor<
        s64[4]
        [0, 1, 0, 1]
      >

      iex> Scholar.Utils.normalize_labels(Nx.tensor([1, 2, 2, 0, 1, 2, 0]), num_classes: 3)
      #Nx.Tensor<
        s64[7]
        [0, 1, 1, 2, 0, 1, 2]
      >

  """
  defn normalize_labels(labels, opts \\ []) do
    opts = keyword!(opts, [:num_classes])

    num_classes = opts[:num_classes]

    classes = Nx.iota({num_classes})

    class_min_idx =
      Nx.equal(
        Nx.new_axis(classes, 1),
        Nx.new_axis(labels, 0)
      )
      |> Nx.argmax(axis: 1)

    mapping = to_consecutive(class_min_idx)

    Nx.take(mapping, labels)
  end

  # This is effectively a double argsort, just optimised
  defnp to_consecutive(x) do
    x = Nx.argsort(x)
    Nx.indexed_add(Nx.broadcast(0, x), Nx.new_axis(x, -1), Nx.iota(x))
  end
end
```

jonatanklosko commented on lib/scholar/cluster/k_means.ex:
```diff
@@ -0,0 +1,337 @@
+      Use an int to make the randomness deterministic.
+      The argument `nil` is special and means that the seed is not set. Defaults to `nil`.
+
+    * `:weights` - The weights for each observation in x. If equals to `nil`, all observations
+      are assigned equal weight.
```
It seems like this should be a tensor argument, as we convert it to a tensor anyway. Since it is one weight per each data point in `x`, the weights may be huge and we don't want to embed all of that as a constant into the computation.

One option is:

```elixir
defn fit(x, opts \\ []) do
  weights = Nx.broadcast(1, ...)
  fit(x, weights, opts)
end

defn fit(x, weights, opts) do
  ...
end
```

@polvalente @josevalim thoughts?

jonatanklosko replied:
FTR we can support tensors in options by having `fit/2` as a transform and let it call the `fit/3` defn. We will worry about this later.

jonatanklosko commented on lib/scholar/cluster/k_means.ex:
```diff
@@ -0,0 +1,337 @@
+    * `:weights` - The weights for each observation in x. If equals to `nil`, all observations
+      are assigned equal weight.
+
+    * `:init` - Method for initialization (Defaults to `:k_means_plus_plus`):
+
+        * `:k_means_plus_plus` -  selects initial cluster centroids using sampling based on an empirical probability
+         distribution of the points’ contribution to the overall inertia.
+         This technique speeds up convergence, and is theoretically proven to be O(log(k))-optimal.
+
+        * `:random` - choose :num_clusters observations (rows) at random from data for the initial centroids.
```
```suggestion
    * `:init` - Method for centroid initialization, either of:

        * `:k_means_plus_plus` (default) -  selects initial cluster centroids using sampling based on
          an empirical probability distribution of the points’ contribution to the overall inertia.
         This technique speeds up convergence, and is theoretically proven to be O(log(k))-optimal.

        * `:random` - choose `:num_clusters` observations (rows) at random from data for the initial centroids.
```

jonatanklosko commented on test/scholar/cluster/k_means_test.exs:
```diff
@@ -0,0 +1,171 @@
+      )
+
+    {oredered_clusters, ordered_labels} = sort_clusters(model.clusters, model.labels)
+    model = %{model | :labels => ordered_labels}
+    model = %{model | :clusters => oredered_clusters}
```
```suggestion
    model = %{model | labels: ordered_labels, clusters: oredered_clusters}
```

Also, let's do that in `sort_clusters`, so it's shared across tests :)

msluszniak commented on lib/scholar/cluster/k_means.ex:
```diff
@@ -0,0 +1,323 @@
+             inertia = Nx.broadcast(0.0, {num_runs}), weights, broadcast_weights, broadcast_x,
+             centroids, nearest_centroids = Nx.broadcast(-1, {num_runs, num_samples})},
+            i < opts[:max_iterations] and
+              Nx.all(distance > opts[:tol]) do
```
I missed that, let me fix this

msluszniak commented:
I added a livebook with a few examples of KMeans usage. In the livebook, I implemented the Silhouette module which, in my opinion, might be added to metrics in scholar, what do you think?

josevalim commented:
> In the livebook, I implemented the Silhouette module which, in my opinion, might be added to metrics in scholar, what do you think?

Do we have similar metrics in sklearn? If we do, we should go ahead and add them here as well. :+1:

msluszniak commented:
> 

https://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_score.html and https://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_samples.html#sklearn.metrics.silhouette_samples so yes, they have these functionalities

josevalim commented:
@msluszniak fantastic, so let’s add them too!!!

hugobarauna commented on notebooks/k_means.livemd:
```diff
@@ -0,0 +1,559 @@
+    {:nx, github: "elixir-nx/nx", sparse: "nx", override: true},
+    {:exla, "~> 0.3"},
+    {:explorer, "~> 0.2.0"},
+    {:scholar, path: "/home/mateusz/Dashbit/scholar"},
```
I was trying to run this notebook on my machine and got a problem with this path.

Since we're using path option for a mix dependency, assuming that the person running this notebook is running it within the scholar git repo, maybe we could change that hardcoded path to a relative one, like this:

```elixir
{:scholar, path: Path.join(__DIR__, "..")},
```

hugobarauna commented on notebooks/k_means.livemd:
````diff
@@ -0,0 +1,559 @@
+Let us start with loading the referral image.
+
+```elixir
+{:ok, img} = StbImage.read_file("../scholar/test/data/taj_mahal.jpg")
````
```suggestion
{:ok, img} = StbImage.read_file(Path.join(__DIR__, "../test/data/taj_mahal.jpg"))
```

jonatanklosko replied:
@msluszniak are there reliable locations we can download the image and the csv file from? Ideally we would download them with `req`, so the notebook runs without depending on local files.

msluszniak replied:
The CSV data is available [here](https://cam-orl.co.uk/facedatabase.html), with the image I think I should change it, cause I cannot find the usage rights.

msluszniak replied:
@jonatanklosko maybe it will be better to add the Olivetti dataset to [Scidata](https://github.com/elixir-nx/scidata) module. What do you think?

jonatanklosko replied:
Sounds good to me!

josevalim replied:
@msluszniak yes, please do send a PR to Scidata. It also has some image datasets if you would prefer to use that instead. :)

msluszniak replied:
@josevalim unfortunately there is a problem with Olivetti. The data source (sci-kit uses it; [link](https://github.com/scikit-learn/scikit-learn/blob/36958fb24/sklearn/datasets/_olivetti_faces.py#L32)) is saved as a .mat file. It would be problematic to parse this file into CSV. I think that I use fashionmnist instead.

josevalim replied:
Ah, if fashionmnist can work that's great!

hugobarauna commented on notebooks/k_means.livemd:
````diff
@@ -0,0 +1,559 @@
+Firstly, load the data and cast it into Nx tensors.
+
+```elixir
+olivetti_data = Explorer.DataFrame.from_csv!("../scholar/test/data/olivetti_data.csv")
````
```suggestion
olivetti_data = Explorer.DataFrame.from_csv!(Path.join(__DIR__, "../test/data/olivetti_data.csv"))
```

hugobarauna commented on notebooks/k_means.livemd:
````diff
@@ -0,0 +1,559 @@
+
+```elixir
+olivetti_data = Explorer.DataFrame.from_csv!("../scholar/test/data/olivetti_data.csv")
+olivetti_target = Explorer.DataFrame.from_csv!("../scholar/test/data/olivetti_target.csv")
````
```suggestion
olivetti_target = Explorer.DataFrame.from_csv!(Path.join(__DIR__, "../test/data/olivetti_target.csv"))
```

josevalim commented on lib/scholar/cluster/k_means.ex:
```diff
@@ -0,0 +1,308 @@
+  defstruct [:num_iterations, :clusters, :inertia, :labels]
+
+  @doc """
+  Fits a K-Means model for sample inputs `x`
+
```
```suggestion
  Fits a K-Means model for sample inputs `x`.
```

josevalim reviewed (approved):
This is awesome @msluszniak, excellent job! :)

I have approved and someone else can merge it if they also agree. :)

jonatanklosko reviewed (approved):
Beautiful work with the notebook examples! I've just pushed some small changes :)
